test/distributed/_composable/test_composability/test_3d_composability.py

```diff
@@ -0,0 +1,565 @@
+
+        def _build_stage(
+            stage_idx, start_layer, stop_layer, is_first=False, is_last=False
+        ):
+            model = copy.deepcopy(whole_model)
+            if not is_first:
+                model.tok_embeddings = None
+
+            start_idx = int(start_layer.split(".")[-1]) if start_layer else 0
+            stop_idx = int(stop_layer.split(".")[-1]) if stop_layer else num_layers
+            model = model[start_idx:stop_idx]
+
+            if not is_last:
+                model.norm = None
+                model.output = None
+
+            stage = PipelineStage(
+                model,
+                stage_idx,
+                num_stages,
+                device,
+                group=pp_mesh.get_group("pp"),
+            )
+            return stage, model
+
+        num_stages = len(splits) + 1
+        stage_idx = pp_rank
+
+        stages = []
+        models = []
+        stages_per_rank = num_stages // pp_size
+
+        for s in range(stages_per_rank):
+            stage_idx = pp_rank + s * pp_size
+            start_layer = splits[stage_idx - 1] if stage_idx > 0 else None
+            stop_layer = splits[stage_idx] if stage_idx < num_stages - 1 else None
+            stage, model_chunk = _build_stage(
+                stage_idx,
+                start_layer,
+                stop_layer,
+                is_first=stage_idx == 0,
+                is_last=stage_idx == num_stages - 1,
+            )
+            stages.append(stage)
+            models.append(model_chunk)
+        return stages, models
+
+    def generate_split_points(self, schedule_class, pp_dim):
+        schedule_class = get_schedule_class(schedule_class)
+        if issubclass(schedule_class, PipelineScheduleSingle):
+            num_stages_per_rank = 1
+        elif issubclass(schedule_class, PipelineScheduleMulti):
+            # Multi-stage schedules support more than 2 stages per rank, but this is the default if
+            # no pipeline split is specified
+            num_stages_per_rank = 2
+        else:
+            raise ValueError("Unsupported pipeline schedule")
+        total_stages = pp_dim * num_stages_per_rank
+
+        if total_stages > num_layers:
+            raise ValueError("Total stages cannot be greater than the number of layers")
+
+        base_interval = num_layers // total_stages
+        extra_layers = num_layers % total_stages
+
+        splits = []
+        current_layer = 0
+        for i in range(total_stages - 1):
+            if i == 0:
+                current_layer += base_interval
+            else:
+                # Middle stages get an extra layer if there are any remaining
+                if extra_layers > 0:
+                    current_layer += base_interval + 1
+                    extra_layers -= 1
+                else:
+                    current_layer += base_interval
+            splits.append("layers." + str(current_layer))
+        return splits
+
+    def apply_tp(
```

Comment:
it feels like apply_tp is not really doing anything. It will not find any 'tok_embeddings' or 'norm' or 'output' in the MLPModule. 

For both apply_tp and apply_fsdp, the important part is NOT to use the same helpers as torchtitan, but to do something that works correctly.